### ISSUE TITLE
Published file structure concept

### ISSUE THREAD

**Issue by ljkart** [pyblish/pyblish-base#9]
By referencing a lot of pipeline structures and from experience the folder structure is something which is specific to each company depends on their different requirements.

So what I would like to suggest is that we can make the folder structure configurable which is flexible and can focus on how data is being processed.
For the time being we can take one default structure and start work on it.

Eg structure:

For Asset

```
/published
  /base_type(char, props, sets, rig)
    /asset_type(forest)
     /asset_name(forest_01)
       /department
    /version
     /scenes
       file.ma
     /images
```

For Shot

```
/published
  /seq
    /shot
      /department
    /version
         /scenes
          file.ma
         /cache
         /images
         /movie
```

  is that fine for you?

**Comment by mottosso** [comment]
This is an important point and one I think most will have conflicting views upon. 

Configurable defaults are one way to deal with it, but I think the issue with any defaults are, what should get included? For example, some might not have a concept of departments and instead work within a single, smaller group, some may not use "types" for their asset but rather work on a single "type", some may have an additional concept of "levels" which would have to be discarded for all film and commercial houses, and so on.

With too many variables, configuration can get quite nasty - foremost for those who won't be needing some of the more complex options.

At the moment, we've got publishes landing directly in the Maya project directory:

``` bash
asset/
    maya/
        scenes/
            my_scene_v001.ma
        publishes/
            v001
            v002
```

This is assuming users work _within_ an asset, and if so there shouldn't be any need to specify the asset again, and no need to specify what type it is. The name and type would already be present within the hierarchy.

``` bash
 # "Peter", and "model" already in the hierarchy
/projects/spiderman/Peter/models/body/maya/publishes/v001
```

### Monolithic project

The other concern is when people aren't working within the asset. Some may for example work like this:

``` bash
/projects/spiderman/maya/scenes/models/Peter/my_scene_v001.ma
```

And from there, publishing under the Maya project directory won't work as we wouldn't be able to distinguish between assets once there:

``` bash
spiderman/
    maya/
        scenes/
            Peter/
                rig/
                   my_scene_v001.ma
                model/
                   my_scene_v001.ma
        publishes/
            v001  # Is this version 1 of Peter the "model" or the "rig"?
```

### Suggestion

My suggestion would be to let publishes land directly in the Maya project directory, and to add an attribute in Maya to override it.

``` bash
Peter.dest = "Peter/model"
```

This way, those working _within_ the asset won't have to worry and those working in a monolithic project could append this property to each of their assets.

``` python
custom_destination = cmds.getAttr(node + ".dest")
destination = os.path.join(workspace, custom_destination)
```

### Further customisations

For those looking for more customisations, I'd suggest having Publish output/emit the path to wherever the publish ends up and let external processes handle the absolute positioning needed for each individual studio.

Thoughts?

**Comment by ljkart** [comment]
That was a good one!
the default one has to something very basic which is compatible to everyone.
adding an destination attribute to override can solve the customization.

**Comment by mottosso** [comment]
Excellent, glad you agree! Would you care to prepare a scene with this in action, so we can see how and if it works out?

**Comment by ljkart** [comment]
I have idea in my mind. what about having a template system for file structure to configure?
eg: /sequence/{seq}/{shot}/{dept}/maya/publish/{version}/scene/{file}
can keep a key in our template library to say **maya_publish_path** and define the path.

**Comment by mottosso** [comment]
Sure, let's try that!

**Comment by ljkart** [comment]
## **Using Template System**

Basically publish function process an input file and saves the output.

```
                         Application
                   _____________________
                  |                     |
----input-------> |       Publish       | --------output-------->
                  |_____________________|

```

The input file can be an asset or a shot etc, which stays somewhere in the centralized 
location in the drive and we load that file into the scene using some utility like 
asset/shot loader. I am pretty much sure that most of the studios follows some sort of 
folder structure to save and load data. for example,

eg: 
&nbsp;&nbsp;$Root/sequences/seq100/shot200/lighting/v001/maya/scenes/file_name.ma - sequences
&nbsp;&nbsp;$Root/assets/characters/piggy/modeling/v002/maya/scenes/file_name.ma - Assets

or something else.
Having a folder structure good enough to build a template system for any pipeline.
Here I'm going to define a template system for publish.

so first of all, we will have to have an API to do that. 
Note : will have to set "assets" and "sequences" template path before we start to create one for publish

```
>>>import publish
>>>publish.write.project_root("/projects/my_project")   # setup the project root location
>>>seq_template_obj = publish.template.create(name="sequences_path", 
                          template_type="sequences", 
                          path="/sequences/{sequence}/{shot}/{stage}/v{version}/{type}/{subtype}/{file_name}.v{version}.ma")
```

And we can create template path for publish

```
>>>template_obj = publish.template.create(name="maya_shot_publish", 
                      template_type="sequences", 
                      app="maya", 
                      path="/published/{sequence}/{shot}/{stage}/v{version}/{type}/{subtype}/{file_name}.v{version}.ma")
>>>print template_obj
template_type:'sequences', 'maya_shot_publish':'/projects/my_project/published/{sequence}/{shot}/{stage}/v{version}/{type}/{subtype}/{sequence}_{shot}_{stage}_{name}.v{version}.ma', app:'maya'
>>>template_obj.read.template_path()    # get the template path
'/projects/my_project/published/{sequence}/{shot}/{stage}/v{version}/{type}/{subtype}/{sequence}_{shot}_{stage}_{name}.v{version}.ma'

```

here we are creating a "maya_shot_publish" (app=maya) template for sequences.

"/published/{sequence}/{shot}/{stage}/v{version}/{type}/{subtype}/{file_name}.v{version}.ma"
&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|
&nbsp;&nbsp;&nbsp;&nbsp;v&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;v
&nbsp;&nbsp;&nbsp;static&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;dynamic
&nbsp;&nbsp;&nbsp;type&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;type

static-type  : will create the folder as it is.
dynamic-type : will replace with the actual value.

This can easily extracts every fields for every given path.
for eg: if we have loaded a shot "/projects/my_project/sequences/seq100/shot200/lighting/v001/maya/scenes/file_name.ma" in maya

then inside Maya we can do the following,

```
>>>import publish
>>>pub_template_obj = publish.template.read(name="maya_shot_publish" )
>>>pub_template_obj.read.publish_path()
'/projects/my_project/published/seq100/shot200/lighting/v002/maya/scenes/file_name.ma'
>>>template_obj.read.fields()
{'sequences': 'seq100',
'shot': 'shot200',
'stage': 'lighting',
'type', 'maya',
'subtype': 'scenes',
'file': 'file_name.ma'}

```

Then we can easily store this value either as metadata inside our publishable sets or we can retrieve this at the time of publish.

But how do we store all these details ,
for that we have to have a publish configuration system. so when we install publish tool, 
it automatically creates a publish configuration system for you.
for eg:
&nbsp;&nbsp;/publish
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;/projects
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;/my_project
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;/my_project.type
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;/maya
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;/maya.type
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;/sequences
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;/sequence
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;/sequence.json
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;/shot
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;/shot.json
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;/stage
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;/stage.json
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;.
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;.
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;.
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;/assets
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;.
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;.
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;/published

This is a concept folder structure for publish tool, it is bieng created when you do initial configuration for publish tool for every project.
You have noticed that some of the folders have a JSON file inside, this is for storing metadata for that file, if it is present, then the folder 
considered to be a dynamic type  which can be easily renamed by our template api and whoever doesn't have the JSON file, considered to be a static file which is just copied
to destination.

The JSON file mainly contain info like type of the file - eg:sequence, shot, stage, versions etc. Even you can add more information to the file as metadata as you need.
why this folder structure is helpfull because if you are going to setup publish tool for a new project then you want to maintain old project configuration 
then you can simply say,

```
>>>publish.template.load_from(path)
```

which will copy the configuration for you from the old project.

I know this is kinda lengthy one, but I would say this is only a concept for setting up publish file structure and even subjected to many changes as per the requrement.

**Comment by mottosso** [comment]
Excellent write-up Liju! Thanks for taking the time to make it pretty with Markdown too. :) Before we go further, I'd like to touch on just two aspects.
1. Are we all in agreement on where the responsibilities of Publish ends? This is how I envision files to travel, let me know if you have got the same idea:
   1. 1. After validation, files are exported on to a temporary space on the local computer; so as to not write   directly from e.g. Maya onto a network share (if you're wondering why that is necessary, let's go  through it here as well)
   
   ``` bash
   # Example destination from Maya
   $ /tmp/5424ftgsd/my_temporary_publish.obj
   ```
   1. 2. Once complete, files are moved onto a location relative the working file; the working file is assumed to be accessible by other pipeline tools.
   
   ``` bash
   # Example destination from temporary location
   $ /spiderman/Peter/mottosso/maya/published/140803_0759_1243.mb
   # Note that we can't safely assume any versioning here, as
   # there may be multiple artists working on the same asset,
   # each with his own "maya" working directory.
   ```
   1. 3. Publish is done. An event/callback is triggered to further organise the output. This is where your template system would come into play, @ljkart but in my opinion it should be optional and mainly for those who don't already have a system for organising their files. It would then be capable of determining an absolute version of published files.
   
   ``` bash
   # Example output from Publish location
   $ /spiderman/Peter/published/v054.ma
   ```
   
   I've tried nailing down exactly where we start and stop here, if you find anything different from your understanding, let's go through it here.
   https://github.com/abstractfactory/publish/wiki/What-is-publishing
2. The next thing is about existing template systems, something we could either learn from or extend upon directly. Here are the ones I can think of off the top of my head, let's talk about why these are or aren't a good fit.
   1. [Shotgun template system](https://github.com/shotgunsoftware/tk-core/blob/master/python/tank/template.py)
   2. [Katana asset management](http://thefoundry.s3.amazonaws.com/products/katana/releases/1.6v2/Katana_1.6v2_TechnicalGuide.pdf)
   3. [bsemantic](https://github.com/Byron/bcore/tree/master/src/python/bsemantic)
   
   Shotgun has open sourced parts of their system, but I'm unaware of whether this includes the full parser or not. Bsemantic on the other hand is by a friend of mine, Sebastian, and does but lacks documentation. Katana has something like it, but I'm unsure of exactly how it works.

**Comment by mottosso** [comment]
@BigRoy mentioned another alternative [here](http://tech-artists.org/forum/showthread.php?4874-Open-Source-Publishing-Development&p=25154&viewfull=1#post25154) which I found intriguing; use of a web-template engine e.g. [Jinja2](http://jinja.pocoo.org/docs/)

Web is probably the source of templates in the first place, so it may be helpful to study its history of pros and cons.

**Comment by ljkart** [comment]
Hi Marc,
1. Why we have to save the file in temp if the user has a workspace?

> (2.) Once complete, files are moved onto a location relative the working file; the working file is assumed to be accessible by other pipeline tools.

```
# local publish
$ /spiderman/Peter/mottosso/maya/published/140803_0759_1243.mb
```

```
# final publish
$ /spiderman/Peter/published/v054.ma
```

I am just summarizing your point, correct me if I am wrong,
- As you said we have to publish file twice , one is local publish(it can be a manual file save) and other is actual publish.
-  our tool will do the actual publish.

> in my opinion it should be optional and mainly for those who don't already have a system for organising their files. It would then be capable of determining an absolute version of published files.

Even if they have a system to organize files, how do our tool know where we have to publish, we have to set the output location right?

**Comment by mottosso** [comment]
Hi Liju,

(I added parenthesis around the "2" in your post above due to Markdown trying to be smart and assume you start lists with a one and replace it)

> As you said we have to publish file twice , one is local publish(it can be a manual file save) and other is actual publish.

Actually, three times. 
1. Locally
2. Centrally
3. Final

Final being the designated location for assets studio-wide, which is outside of our control.
1. to the local drive
   
   ``` bash
   Maya --------> C:/tmp/my_temporary_publish.obj
   ```
2. to a directory relative the users working file (must be central):
   
   ``` bash
   # Assuming the users working file is at '/spiderman/Peter/mottosso/maya'
   c:/tmp --> /spiderman/Peter/mottosso/maya/published/140803_0759_1243.mb
   ```
   
   This is where Publish ends. It will know of this directory because of the user's working file. The "alghorithm" would simply be:
   
   ``` python
   published_path = os.path.join(cmds.workspace(rootDir=True, query=True), 'published')
   ```
3. To the final destination; either using the template system, or existing mechanism

``` python
os.move(published_files, final_destination)
```

> Even if they have a system to organize files, how do our tool know where we have to publish, we have to set the output location right?

The way I'm imagining them organising these files is either directly, via callback or message.

``` python
import our_tool
import publish

# Directly
published_path = publish.publish()
final_path = our_tool.place_publish(published_path)

# Callback
publish.published.connect(our_tool.place_publish)

# Signal/message
our_tool.listen('192.0.0.125:4040', our_tool.place_publish)
```

## Publish template system

To clarify however the usefulness of providing the mechanism ourselves; some, if not most, will most likely not be using any system for paths (i.e. smaller shops). If we can provide a starting point for them, and completely govern the placing of assets, I think Publish would be made a much more attractive option for those without a pipeline or pipeline department. Those are the ones I'm imagining this template system would work for.

Thoughts?

**Comment by BigRoy** [comment]
I would at least allow the studio to directly configure _publish_ to save to their preferred final position. If a big team keeps moving files around on a server that can easily be of insane sizes (Alembic caches go up pretty quickly in size!) will only drag down the workflow.

The 'after-export' callback is very nice so other tools can pick up or even allow a new _publish_ sequence to be picked up, like a validation of the output of sorts. But I think just being able to publish.publish(template, keys=myKeys)

That could either use a native templating system (replacing certain dynamic parts with other keys) or a static path to be set.

**Comment by mottosso** [comment]
> I would at least allow the studio to directly configure publish to save to their preferred final position. If a big team keeps moving files around on a server that can easily be of insane sizes (Alembic caches go up pretty quickly in size!) will only drag down the workflow.

That's a good idea, we could allow to either hook into a publish _after_ extraction, or to customise the _during_ of extraction; essentially re-directing step 2 onto a final destination.

> But I think just being able to publish.publish(template, keys=myKeys)

And this could be a way of doing just that. Let's run with this exact interface and see where it takes us.

**Comment by mottosso** [comment]
Mentioning this here from #18, about templates and final destination of assets.
- [FTrack Location API](http://www.fxguide.com/featured/ftrack-debuts-v3-at-siggraph/)

**Comment by achayan** [comment]
I can work on this issue.

**Comment by mottosso** [comment]
Excellent. Do you have any thoughts about what's been said here so far?

**Comment by achayan** [comment]
I think this is already in a good shape and it will work perfect.

**Comment by mottosso** [comment]
As the requirements have become more clear about published destinations, if there is anything to add, either open up a new issue or a conversation in the usergroup and we'll continue from there.
